File: elixir/tenpin_test.exs
ExUnit.start()

defmodule TenpinTest do
  Code.require_file "tenpin.exs", __DIR__

  use ExUnit.Case
  alias Bowling.Tenpin, as: Tenpin

  test "initial score" do
    assert Tenpin.score([]) == 0
  end

  test "complete open frame" do
    assert Tenpin.score([1,1]) == 2
  end

  test "multiple complete open frames" do
    assert Tenpin.score([1,1,1,1,1,1,1,1,1,1]) == 10
  end

  test "incomplete open frame" do
    assert Tenpin.score([1]) == 0
  end

  test "multiple open frames with last incomplete" do
    assert Tenpin.score([1,1,1,1,1]) == 4
  end

  test "all gutter balls" do
    rolls = [0,0,0,0,0,0,0,0,0,0]
    assert Tenpin.score(rolls) == 0
  end

  test "complete strike followed by complete open frame" do
    assert Tenpin.score([10,1,1]) == 14
  end

  test "complete strike followed by incomplete spare" do
    assert Tenpin.score([10,5,5]) == 20
  end

  test "complete strike followed by complete spare" do
    assert Tenpin.score([10,5,5,1]) == 31
  end

  test "incomplete strike, no bonus rolls" do
    assert Tenpin.score([10]) == 0
  end

  test "incomplete strike, one bonus roll" do
    assert Tenpin.score([10,1]) == 0
  end

  test "perfect game" do
    rolls = [10,10,10,10,10,10,10,10,10,10,10,10]
    assert Tenpin.score(rolls) == 300
  end

  test "complete spare" do
    assert Tenpin.score([3,7,4]) == 14
  end

  test "complete spare followed by complete open frame" do
    assert Tenpin.score([3,7,4,1]) == 19
  end

  test "complete spare followed by strike with no bonus rolls" do
    assert Tenpin.score([3,7,10]) == 20
  end

  test "complete spare followed by strike with one bonus roll" do
    assert Tenpin.score([3,7,10,1]) == 20
  end

  test "all fives" do
    rolls = [5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5]
    assert Tenpin.score(rolls) == 150
  end
end

File: ruby/tenpin_test.rb
require_relative "tenpin"
require "minitest/autorun"

class TestTenpin < Minitest::Test
  def setup
    @game = Bowling::Tenpin.new
  end

  def test_initial_score
    assert_equal @game.score, 0
  end

  def test_complete_open_frame
    2.times { @game.roll(1) }
    assert_equal 2, @game.score
  end

  def test_multiple_complete_open_frames
    10.times { @game.roll(1) }
    assert_equal 10, @game.score
  end

  def test_incomplete_open_frame
    @game.roll(1)
    assert_equal 0, @game.score
  end

  def test_multiple_open_frames_with_last_incomplete
    5.times { @game.roll(1) }
    assert_equal 4, @game.score
  end

  def test_all_gutter_balls
    20.times { @game.roll(0) }
    assert_equal 0, @game.score
  end

  def test_complete_strike_followed_by_complete_open_frame
    [10,1,1].each { |pinfall| @game.roll(pinfall) }
    assert_equal 14, @game.score
  end
end

class TestFrame < Minitest::Test
  def setup
    @frame = Bowling::Frame.new
  end

  def test_incomplete_open_frame
    @frame.roll(1)
    assert_equal 0, @frame.score
  end

  def test_complete_open_frame
    2.times { @frame.roll(1) }
    assert_equal 2, @frame.score
  end

  def test_strike
    @frame.roll(10)
    assert @frame.strike?
  end

  def test_spare
    2.times { @frame.roll(5) }
    assert @frame.spare?
  end

  def test_open
    2.times { @frame.roll(1) }
    assert @frame.open?
  end

  def test_complete_strike
    [10,1,1].each { |pinfall| @frame.roll(pinfall) }
    assert_kind_of Bowling::CompleteState, @frame.state
  end

  def test_complete_spare
    [5,5,1].each { |pinfall| @frame.roll(pinfall) }
    assert_kind_of Bowling::CompleteState, @frame.state
  end

  def test_complete_open
    2.times { @frame.roll(1) }
    assert_kind_of Bowling::CompleteState, @frame.state
  end

  def test_incomplete_strike
    [10,1].each { |pinfall| @frame.roll(pinfall) }
    assert_kind_of Bowling::BonusState, @frame.state
  end

  def test_incomplete_spare
    [8,2].each { |pinfall| @frame.roll(pinfall) }
    assert_kind_of Bowling::BonusState, @frame.state
 end

  def test_incomplete_open
    @frame.roll(9)
    assert_kind_of Bowling::PendingState, @frame.state
  end
end
